module: refator ESM loader for adding future synchronous hooks

joyeecheung opened the pull request:
This lays the foundation for supporting synchronous hooks proposed in https://github.com/nodejs/loaders/pull/198 for ESM.

- Corrects and adds several JSDoc comments for internal functions of the ESM loader, as well as explaining how require() for import CJS work in the special resolve/load paths. This doesn't consolidate it with import in require(esm) yet due to caching differences, which is left as a TODO.
- The moduleProvider passed into ModuleJob is replaced as moduleOrModulePromise, we call the translators directly in the ESM loader and verify it right after loading for clarity.
- Reuse a few refactored out helpers for require(esm) in getModuleJobForRequire().

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

joyeecheung commented on lib/internal/modules/esm/loader.js:
```diff
@@ -370,23 +373,16 @@ class ModuleLoader {
-    const wrap = FunctionPrototypeCall(translator, this, responseURL, source, isMain);
-    assert(wrap instanceof ModuleWrap);  // No asynchronous translators should be called.
+    const wrap = this.#translate(url, finalFormat, source, isMain);
+    assert(wrap instanceof ModuleWrap, `Translator used for require(${url}) should not be async`);
```
The assertions for translators could be removed if we just switch to `compiled = new WebAssembly.Module(source)` in the wasm translator, though I am not sure if anyone still wants to keep the imported wasm asynchronously compiled (import 'wasm' is not really all that useful after https://github.com/WebAssembly/esm-integration/blob/main/proposals/esm-integration/README.md is shipped by V8 anyway)

RedYetiDev commented:
nit: refator->refactor in commit

mcollina reviewed (approved):
lgtm

jasnell reviewed (approved):
Typo in the commit message but otherwise LGTM

joyeecheung commented:
Fixed the typo commit message - also I noticed there are a few typos in the return type of `getModuleJobForImport` and `getModuleJobForRequireInImportedCJS` (deleted the wrong `Promise<` and `>` after copy-pasting somehow) so I fixed that as well.

joyeecheung commented:
Rebased to resolve the merge conflicts.

joyeecheung commented:
@mcollina @Qard @jasnell this needs a re-approval after the rebase. Can you take a look again? Thanks

laverdet commented:
@joyeecheung was it intentional to remove `responseURL`?

```
$ cat loader.mjs 
export async function load(urlString, context, nextLoad) {
	const result = await nextLoad(urlString, context);
	return {
		...result,
		responseURL: 'test://',
	};
}

$ cat register.mjs 
import { register } from 'node:module';

register('./loader.mjs', {
	parentURL: import.meta.url,
});

$ cat main.mjs 
console.log(import.meta.url);

$ node -v
v22.4.1

$ node --import ./register.mjs main.mjs                                                         
test://

$ [...]
$ node -v
v22.10.0

$ node --import ./register.mjs main.mjs  
file:///Users/marcel/tmp/main.mjs
```

This feature was useful because it allows you to control `import.meta.url`. I used it pretty significantly to hide implementation details in my hot reloader [dynohot](https://github.com/braidnetworks/dynohot). The resolve hook would resolve something like `"hot:module?url=file:///file.mjs"` and the load hook knows how to unwrap that parameter and load the underlying resource. We don't want to affect the result of `import.meta` at runtime though.

joyeecheung commented:
> was it intentional to remove responseURL?

It was not intended to break any public API, though it seems responseURL was never properly documented nor tested. Do you want to submit a PR to add it back with proper tests and documentation?
